File: backend/lib/athasha_runners/modbus.ex
defmodule Athasha.Runner.Modbus do
  alias Modbus.Master
  alias Athasha.Slave
  alias Athasha.Raise
  alias Athasha.PubSub
  alias Athasha.Number
  alias Athasha.Item
  alias Athasha.Bus
  @status 1000

  def run(item) do
    id = item.id
    config = item.config
    setts = config["setts"]
    trans = setts["trans"]
    proto = setts["proto"]
    host = setts["host"]
    tty = setts["tty"]
    dbpsb = setts["dbpsb"]
    port = String.to_integer(setts["port"])
    speed = String.to_integer(setts["speed"])
    period = String.to_integer(setts["period"])
    password = setts["password"]

    inputs =
      Enum.map(config["inputs"], fn input ->
        slave = String.to_integer(input["slave"])
        address = String.to_integer(input["address"]) - 1
        code = input["code"]
        name = input["name"]
        factor = Number.parse_float!(input["factor"])
        offset = Number.parse_float!(input["offset"])

        %{
          id: "#{id} #{name}",
          name: name,
          getter: getter(code, slave, address),
          calib: Number.calibrator(factor, offset)
        }
      end)

    outputs =
      Enum.map(config["outputs"], fn output ->
        slave = String.to_integer(output["slave"])
        address = String.to_integer(output["address"]) - 1
        code = output["code"]
        name = output["name"]
        factor = Number.parse_float!(output["factor"])
        offset = Number.parse_float!(output["offset"])

        %{
          id: "#{id} #{name}",
          name: name,
          setter: setter(code, slave, address),
          calib: Number.calibrator(factor, offset),
          reverse: Number.reversed(factor, offset)
        }
      end)

    config = %{
      item: Item.head(item),
      trans: trans,
      proto: proto,
      host: host,
      port: port,
      tty: tty,
      speed: speed,
      dbpsb: dbpsb,
      period: period,
      inputs: inputs,
      outputs: outputs
    }

    case connect_master(config) do
      {:ok, master} ->
        # avoid registering duplicates
        Enum.reduce(inputs, %{}, fn input, map ->
          iid = input.id

          if !Map.has_key?(map, iid) do
            PubSub.Input.register!(id, iid, input.name)
          end

          Map.put(map, iid, iid)
        end)

        # avoid registering duplicates
        Enum.reduce(outputs, %{}, fn output, map ->
          oid = output.id

          if !Map.has_key?(map, oid) do
            PubSub.Output.register!(id, oid, output.name)
          end

          Map.put(map, oid, oid)
        end)

        names = Enum.map(inputs, & &1.name)
        PubSub.Input.reg_names!(id, names)
        names = Enum.map(outputs, & &1.name)
        PubSub.Output.reg_names!(id, names)
        Enum.each(outputs, fn output -> Bus.register!({:write, output.id}) end)
        run_once(item, config, master, %{})
        PubSub.Status.update!(item, :success, "Connected")
        PubSub.Password.register!(item, password)
        Process.send_after(self(), :status, @status)
        Process.send_after(self(), :once, period)
        run_loop(item, config, master, %{}, period)

      {:error, reason} ->
        Raise.error({:connect, reason})
    end
  end

  defp run_loop(item, config, master, values, period) do
    values = wait_once(item, config, master, values, period)
    run_loop(item, config, master, values, period)
  end

  defp wait_once(item, config, master, values, period) do
    receive do
      :status ->
        PubSub.Status.update!(item, :success, "Running")
        Process.send_after(self(), :status, @status)
        values

      :once ->
        run_once(item, config, master, values)
        Process.send_after(self(), :once, period)
        %{}

      {{:write, id}, _, value} ->
        Map.put(values, id, value)

      other ->
        Raise.error({:receive, other})
    end
  end

  defp run_once(%{id: id}, config, master, values) do
    Enum.each(config.inputs, fn input ->
      case input.getter.(master) do
        {:ok, value} ->
          value = input.calib.(value)
          PubSub.Input.update!(id, input.id, input.name, value)

        {:error, reason} ->
          Raise.error({:input, reason, input.id})
      end
    end)

    Enum.each(config.outputs, fn output ->
      value = values[output.id]

      if value != nil do
        value = output.calib.(value)

        case output.setter.(master, value) do
          {:ok, value} ->
            reversed = output.reverse.(value)
            PubSub.Output.update!(id, output.id, output.name, reversed)

          {:error, reason} ->
            Raise.error({:output, reason, output.id, value})
        end
      end
    end)
  end

  defp getter(code, slave, address) do
    case code do
      # 03
      "01 Coil" ->
        fn master -> read_bit(master, :rc, slave, address) end

      "02 Input" ->
        fn master -> read_bit(master, :ri, slave, address) end

      "03 U16BE" ->
        fn master -> read_register(master, :rhr, slave, address, &Number.r_u16be/1) end

      "03 S16BE" ->
        fn master -> read_register(master, :rhr, slave, address, &Number.r_s16be/1) end

      "03 U16LE" ->
        fn master -> read_register(master, :rhr, slave, address, &Number.r_u16le/1) end

      "03 S16LE" ->
        fn master -> read_register(master, :rhr, slave, address, &Number.r_s16le/1) end

      # F32
      "03 F32BED" ->
        fn master -> read_register2(master, :rhr, slave, address, &Number.r_f32bed/1) end

      "03 F32LED" ->
        fn master -> read_register2(master, :rhr, slave, address, &Number.r_f32led/1) end

      "03 F32BER" ->
        fn master -> read_register2(master, :rhr, slave, address, &Number.r_f32ber/1) end

      "03 F32LER" ->
        fn master -> read_register2(master, :rhr, slave, address, &Number.r_f32ler/1) end

      # S32
      "03 S32BED" ->
        fn master -> read_register2(master, :rhr, slave, address, &Number.r_s32bed/1) end

      "03 S32LED" ->
        fn master -> read_register2(master, :rhr, slave, address, &Number.r_s32led/1) end

      "03 S32BER" ->
        fn master -> read_register2(master, :rhr, slave, address, &Number.r_s32ber/1) end

      "03 S32LER" ->
        fn master -> read_register2(master, :rhr, slave, address, &Number.r_s32ler/1) end

      # U32
      "03 U32BED" ->
        fn master -> read_register2(master, :rhr, slave, address, &Number.r_u32bed/1) end

      "03 U32LED" ->
        fn master -> read_register2(master, :rhr, slave, address, &Number.r_u32led/1) end

      "03 U32BER" ->
        fn master -> read_register2(master, :rhr, slave, address, &Number.r_u32ber/1) end

      "03 U32LER" ->
        fn master -> read_register2(master, :rhr, slave, address, &Number.r_u32ler/1) end

      # 04
      "04 U16BE" ->
        fn master -> read_register(master, :rir, slave, address, &Number.r_u16be/1) end

      "04 S16BE" ->
        fn master -> read_register(master, :rir, slave, address, &Number.r_s16be/1) end

      "04 U16LE" ->
        fn master -> read_register(master, :rir, slave, address, &Number.r_u16le/1) end

      "04 S16LE" ->
        fn master -> read_register(master, :rir, slave, address, &Number.r_s16le/1) end

      # F32
      "04 F32BED" ->
        fn master -> read_register2(master, :rir, slave, address, &Number.r_f32bed/1) end

      "04 F32LED" ->
        fn master -> read_register2(master, :rir, slave, address, &Number.r_f32led/1) end

      "04 F32BER" ->
        fn master -> read_register2(master, :rir, slave, address, &Number.r_f32ber/1) end

      "04 F32LER" ->
        fn master -> read_register2(master, :rir, slave, address, &Number.r_f32ler/1) end

      # S32
      "04 S32BED" ->
        fn master -> read_register2(master, :rir, slave, address, &Number.r_s32bed/1) end

      "04 S32LED" ->
        fn master -> read_register2(master, :rir, slave, address, &Number.r_s32led/1) end

      "04 S32BER" ->
        fn master -> read_register2(master, :rir, slave, address, &Number.r_s32ber/1) end

      "04 S32LER" ->
        fn master -> read_register2(master, :rir, slave, address, &Number.r_s32ler/1) end

      # U32
      "04 U32BED" ->
        fn master -> read_register2(master, :rir, slave, address, &Number.r_u32bed/1) end

      "04 U32LED" ->
        fn master -> read_register2(master, :rir, slave, address, &Number.r_u32led/1) end

      "04 U32BER" ->
        fn master -> read_register2(master, :rir, slave, address, &Number.r_u32ber/1) end

      "04 U32LER" ->
        fn master -> read_register2(master, :rir, slave, address, &Number.r_u32ler/1) end
    end
  end

  defp setter(code, slave, address) do
    case code do
      "05 Coil" ->
        fn master, value -> write_bit(master, :fc, slave, address, value) end

      "06 U16BE" ->
        fn master, value ->
          write_register(master, :phr, slave, address, value, &Number.w_u16be/1)
        end

      "06 S16BE" ->
        fn master, value ->
          write_register(master, :phr, slave, address, value, &Number.w_s16be/1)
        end

      "06 U16LE" ->
        fn master, value ->
          write_register(master, :phr, slave, address, value, &Number.w_u16le/1)
        end

      "06 S16LE" ->
        fn master, value ->
          write_register(master, :phr, slave, address, value, &Number.w_s16le/1)
        end

      # F32
      "16 F32BED" ->
        fn master, value ->
          write_register2(master, :phr, slave, address, value, &Number.w_f32bed/1)
        end

      "16 F32LED" ->
        fn master, value ->
          write_register2(master, :phr, slave, address, value, &Number.w_f32led/1)
        end

      "16 F32BER" ->
        fn master, value ->
          write_register2(master, :phr, slave, address, value, &Number.w_f32ber/1)
        end

      "16 F32LER" ->
        fn master, value ->
          write_register2(master, :phr, slave, address, value, &Number.w_f32ler/1)
        end

      # S32
      "16 S32BED" ->
        fn master, value ->
          write_register2(master, :phr, slave, address, value, &Number.w_s32bed/1)
        end

      "16 S32LED" ->
        fn master, value ->
          write_register2(master, :phr, slave, address, value, &Number.w_s32led/1)
        end

      "16 S32BER" ->
        fn master, value ->
          write_register2(master, :phr, slave, address, value, &Number.w_s32ber/1)
        end

      "16 S32LER" ->
        fn master, value ->
          write_register2(master, :phr, slave, address, value, &Number.w_s32ler/1)
        end

      # U32
      "16 U32BED" ->
        fn master, value ->
          write_register2(master, :phr, slave, address, value, &Number.w_u32bed/1)
        end

      "16 U32LED" ->
        fn master, value ->
          write_register2(master, :phr, slave, address, value, &Number.w_u32led/1)
        end

      "16 U32BER" ->
        fn master, value ->
          write_register2(master, :phr, slave, address, value, &Number.w_u32ber/1)
        end

      "16 U32LER" ->
        fn master, value ->
          write_register2(master, :phr, slave, address, value, &Number.w_u32ler/1)
        end
    end
  end

  defp read_register2(master, code, slave, address, transform) do
    case Master.exec(master, {code, slave, address, 2}) do
      {:ok, [w0, w1]} ->
        value = transform.([w0, w1])
        {:ok, value}

      any ->
        any
    end
  end

  defp read_register(master, code, slave, address, transform) do
    case Master.exec(master, {code, slave, address, 1}) do
      {:ok, [value]} ->
        value = transform.(value)
        {:ok, value}

      any ->
        any
    end
  end

  defp read_bit(master, code, slave, address) do
    case Master.exec(master, {code, slave, address, 1}) do
      {:ok, [value]} -> {:ok, value}
      any -> any
    end
  end

  defp write_bit(master, code, slave, address, value) do
    value = Number.to_bit(value)

    case Master.exec(master, {code, slave, address, value}) do
      :ok -> {:ok, value}
      any -> any
    end
  end

  defp write_register(master, code, slave, address, value, transform) do
    value = transform.(value)

    case Master.exec(master, {code, slave, address, value}) do
      :ok -> {:ok, value}
      any -> any
    end
  end

  defp write_register2(master, code, slave, address, value, transform) do
    case Master.exec(master, {code, slave, address, transform.(value)}) do
      :ok -> {:ok, value}
      any -> any
    end
  end

  defp modbus_proto(%{proto: "TCP"}), do: Modbus.Tcp.Protocol
  defp modbus_proto(%{proto: "RTU"}), do: Modbus.Rtu.Protocol

  defp connect_master(config) do
    proto = modbus_proto(config)

    case config.trans do
      "Socket" ->
        trans = Modbus.Tcp.Transport
        port = Slave.fix_port(config.port)

        case :inet.getaddr(String.to_charlist(config.host), :inet) do
          {:ok, ip} ->
            Master.start_link(trans: trans, proto: proto, ip: ip, port: port)

          any ->
            any
        end

      "Serial" ->
        trans = Modbus.Serial.Transport

        Master.start_link(
          trans: trans,
          proto: proto,
          device: config.tty,
          speed: config.speed,
          config: config.dbpsb
        )
    end
  end
end

File: frontend/src/schemas/Modbus.js
import Check from '../common/Check'
import Comm from '../common/Comm'
import Schema from '../common/Schema'

const inputCodes = [
    "01 Coil",

    "02 Input",

    "03 U16BE",
    "03 S16BE",
    "03 U16LE",
    "03 S16LE",

    "03 F32BED",
    "03 F32LED",
    "03 F32BER",
    "03 F32LER",

    "03 S32BED",
    "03 S32LED",
    "03 S32BER",
    "03 S32LER",

    "03 U32BED",
    "03 U32LED",
    "03 U32BER",
    "03 U32LER",

    "04 U16BE",
    "04 S16BE",
    "04 U16LE",
    "04 S16LE",

    "04 F32BED",
    "04 F32LED",
    "04 F32BER",
    "04 F32LER",

    "04 S32BED",
    "04 S32LED",
    "04 S32BER",
    "04 S32LER",

    "04 U32BED",
    "04 U32LED",
    "04 U32BER",
    "04 U32LER",
]

const outputCodes = [
    "05 Coil",

    "06 U16BE",
    "06 S16BE",
    "06 U16LE",
    "06 S16LE",

    "16 F32BED",
    "16 F32LED",
    "16 F32BER",
    "16 F32LER",

    "16 S32BED",
    "16 S32LED",
    "16 S32BER",
    "16 S32LER",

    "16 U32BED",
    "16 U32LED",
    "16 U32BER",
    "16 U32LER",
]

const protocols = Comm.protocols
const transports = Comm.transports
const serialConfigs = Comm.serialConfigs

function schema() {
    return {
        $type: "object",
        setts: {
            $type: "object",
            proto: {
                value: protocols[0],
                label: "Protocol",
                help: "Select protocol from list",
                check: function (value, label) {
                    Check.inList(value, label, protocols)
                },
            },
            trans: {
                value: transports[0],
                label: "Transport",
                help: "Select transport from list",
                check: function (value, label) {
                    Check.inList(value, label, transports)
                },
            },
            host: {
                value: "127.0.0.1",
                label: "Hostname/IP Address",
                help: "Non empty hostname or IP address",
                check: function (value, label) {
                    Check.notEmpty(value, label)
                },
            },
            port: {
                value: "502",
                label: "TCP Port",
                help: "Non empty integer [0, 65535]",
                check: function (value, label) {
                    Check.isGE(value, label, 0)
                    Check.isLE(value, label, 65535)
                },
            },
            tty: {
                value: "COM1",
                label: "Serial Port Name",
                help: "Select serial port from list"
                    + "\nType begining of name to show completing list"
                    + "\nClick or press ENTER to update list",
                check: function (value, label) {
                    Check.notEmpty(value, label)
                },
            },
            speed: {
                value: "9600",
                label: "Baud Rate",
                help: "Non empty integer > 0",
                check: function (value, label) {
                    Check.isGE(value, label, 1)
                },
            },
            dbpsb: {
                value: "8N1",
                label: "Config",
                help: "Select config from list",
                check: function (value, label) {
                    Check.notEmpty(value, label)
                },
            },
            period: {
                value: "10",
                label: "Period (ms)",
                help: "Non empty integer > 0",
                check: function (value, label) {
                    Check.isInteger(value, label)
                    Check.isGE(value, label, 1)
                },
            },
            password: {
                value: "",
                label: "Access Password",
                help: "Optional access password",
                check: function (value, label) {
                    Check.isString(value, label)
                },
            },
        },
        inputs: {
            $type: "array",
            $value: (value) => [value(0)],
            slave: {
                value: "1",
                header: "Slave Address",
                label: (index) => `Input ${index + 1} Slave Address`,
                help: "Non empty integer [0, 255]",
                check: function (value, label) {
                    Check.isGE(value, label, 0)
                    Check.isLE(value, label, 255)
                },
            },
            code: {
                value: inputCodes[0],
                header: "Function Code",
                label: (index) => `Input ${index + 1} Function Code`,
                help: "Select the function code from list"
                    + "\n01 Read Coil"
                    + "\n02 Read Input"
                    + "\n03 Read Holding Register"
                    + "\n04 Read Input Register"
                    + "\nF Float"
                    + "\nU Unsigned Integer"
                    + "\nS Signed Integer"
                    + "\n16/32 Number of Bits"
                    + "\nBE Big Endian"
                    + "\nLE Little Endian"
                    + "\nD/R Direct/Reversed",
                check: function (value, label) {
                    Check.inList(value, label, inputCodes)
                },
            },
            address: {
                value: (index) => `${index + 1}`,
                header: "Input Address",
                label: (index) => `Input ${index + 1} Address`,
                help: "Non empty integer [1, 65536]",
                check: function (value, label) {
                    Check.isGE(value, label, 1)
                    Check.isLE(value, label, 65536)
                },
            },
            name: {
                value: (index) => `Input ${index + 1}`,
                header: "Input Name",
                label: (index) => `Input ${index + 1} Name`,
                help: "Non empty input name",
                check: function (value, label) {
                    Check.notEmpty(value, label)
                },
            },
            factor: {
                value: "1",
                header: "Value Factor",
                label: (index) => `Input ${index + 1} Factor`,
                help: "Non zero number m in f(x)=m*x+b",
                check: function (value, label) {
                    Check.notZero(value, label)
                },
            },
            offset: {
                value: "0",
                header: "Value Offset",
                label: (index) => `Input ${index + 1} Offset`,
                help: "Non empty number b in f(x)=m*x+b",
                check: function (value, label) {
                    Check.isNumber(value, label)
                },
            },
        },
        outputs: {
            $type: "array",
            $value: [],
            slave: {
                value: "1",
                header: "Slave Address",
                label: (index) => `Output ${index + 1} Slave Address`,
                help: "Non empty integer [0, 255]",
                check: function (value, label) {
                    Check.isGE(value, label, 0)
                    Check.isLE(value, label, 255)
                },
            },
            code: {
                value: outputCodes[0],
                header: "Function Code",
                label: (index) => `Output ${index + 1} Function Code`,
                help: "Select the function code from list"
                    + "\n05 Write Coil"
                    + "\n06 Write Register"
                    + "\n16 Write Registers"
                    + "\nF Float"
                    + "\nU Unsigned Integer"
                    + "\nS Signed Integer"
                    + "\n16/32 Number of Bits"
                    + "\nBE Big Endian"
                    + "\nLE Little Endian"
                    + "\nD/R Direct/Reversed",
                check: function (value, label) {
                    Check.inList(value, label, outputCodes)
                },
            },
            address: {
                value: (index) => `${index + 1}`,
                header: "Output Address",
                label: (index) => `Output ${index + 1} Address`,
                help: "Non empty integer [1, 65536]",
                check: function (value, label) {
                    Check.isGE(value, label, 1)
                    Check.isLE(value, label, 65536)
                },
            },
            name: {
                value: (index) => `Output ${index + 1}`,
                header: "Output Name",
                label: (index) => `Output ${index + 1} Name`,
                help: "Non empty input name",
                check: function (value, label) {
                    Check.notEmpty(value, label)
                },
            },
            factor: {
                value: "1",
                header: "Value Factor",
                label: (index) => `Output ${index + 1} Factor`,
                help: "Non zero number m in f(x)=m*x+b",
                check: function (value, label) {
                    Check.notZero(value, label)
                },
            },
            offset: {
                value: "0",
                header: "Value Offset",
                label: (index) => `Output ${index + 1} Offset`,
                help: "Non empty number b in f(x)=m*x+b",
                check: function (value, label) {
                    Check.isNumber(value, label)
                },
            },
        },
    }
}

function input(index) {
    const $type = "object"
    const prop = { ...schema().inputs, $type }
    return Schema.value(prop, index)
}

function output(index) {
    const $type = "object"
    const prop = { ...schema().outputs, $type }
    return Schema.value(prop, index)
}

export default {
    schema,
    inputCodes,
    outputCodes,
    transports,
    protocols,
    serialConfigs,
    input,
    output,
}
